src/main/java/com/thebois/models/inventory/AbstractInventory.java

```diff
@@ -0,0 +1,85 @@
+package com.thebois.models.inventory;
+
+import java.util.ArrayList;
+import java.util.Optional;
+
+import com.thebois.models.inventory.items.IItem;
+import com.thebois.models.inventory.items.ItemType;
+
+/**
+ * Base implementation for all inventories.
+ */
+public abstract class AbstractInventory {
+
+    private final ArrayList<IItem> items = new ArrayList<>();
+
+    /**
+     * Inputs an item to the inventory.
+     *
+     * @param item The item to be added to the inventory.
+     */
+    public void add(final IItem item) {
+        items.add(item);
+    }
+
+    /**
+     * Removes and returns the first item of the specified type.
+     *
+     * @param itemType The item type to be taken.
+     *
+     * @return The item removed from the inventory.
+     */
+    public Optional<IItem> take(final ItemType itemType) {
+
+        Optional<IItem> item = Optional.empty();
+        if (doesItemTypeExist(itemType)) {
+            final int firstIndexOfItemType = getFirstIndexOf(itemType);
+            item = Optional.ofNullable(items.remove(firstIndexOfItemType));
+        }
+        return item;
+    }
+
+    private int getFirstIndexOf(final ItemType itemType) {
+        int index = -1;
```

Comment:
This line is kind of unecessary since you are doing all the necessary checks before this method is called. 

Might still be worth to keep it for the furture, incase someone doesn't do the necessary checks before.

Reply:
Might have been left from some old version or I just thought it would be good practice. But like we talked about, it might be used insecurely by another method, so this can be seen as defensive programming.

Reply:
Yeah, generally, indexOf-methods return -1 when no item was found. So I agree that it's good to keep in line with common practice.